back/node_watcher/src/tasks/createPreimage.ts

```diff
@@ -135,18 +135,30 @@ const createPreimage: Task<NomidotPreimage[]> = {
           orderBy: 'proposalId_DESC',
         });
 
+        const referenda = await prisma.referendums({
+          where: { preimageHash: h.toString() },
+          orderBy: 'referendumId_DESC',
+        });
+
         const p = proposals[0];
+        const r = referenda[0];
 
         await prisma.createPreimage({
           author: author.toString(),
-          depositAmount: depositAmount?.toString(),
+          depositAmount: depositAmount.toString(),
           hash: h.toString(),
+          metaDescription,
+          method,
+          proposal: p
```

Comment:
question: if a preimage ends up not being connected to a proposal, i.e. `p` was undefined, what's supposed to happen? My guess is that shouldn't happen in the first place and you added the ` : undefined` to satisfy `tsc`?

Reply:
It actually happens in real life, because you "should" actually submit a preimage before submitting a proposal, so that ppl can see what you are proposing.
But you need a deposit for a preimage I think it will be dependant on the code you put onchain. So what happens is that the proposals for `setCode` for instance for a runtime upgrade are often submited after the proposal has been submitted because you don't want to have too much money locked in the deposit, if the referendum doesn't pass.

So in real life, proposal, referendum and preimage don't have a defined order. If you want your proposal to be executed (once it passed), it needs to have a preimage.

On the other side, back to your question, if you submit a preimage, and then never use it for a proposal, it uses storage onchain and there a mechanism so that it can be reaped (after a certain time I imagine, and the reaper will get the deposit back). I don't know the reaping process but there's a preimage event for this.

Reply:
> So what happens is that the ~proposals~ (i guess you mean preimage here?) for setCode for instance for a runtime upgrade are often submited after the proposal has been submitted

so to be clear: you submit proposal, wait for it to go to referendum, then submit a preimage so you can be sure your deposit will locked for as little time as necessary?

also, in this case how can someone decide whether or not to second a proposal if there is no preimage?

also is that deposit decided by the size of the call or preimage, or is it a constant, or something else?

Reply:
> (i guess you mean preimage here?)

Absolutely :)

> so to be clear: you submit proposal, wait for it to go to referendum, then submit a preimage so you can be sure your deposit will locked for as little time as necessary?

Yup, looks like it.
Also I saw Gav ask for someone else to submit the preimage for him because his internet connection was flaky.

> also, in this case how can someone decide whether or not to second a proposal if there is no preimage?

Since the preimage hash is deterministic, you can just check whatever the proposer tells you (off chain, e.g this is a setCode from this branch). The preimage hash is submitted with the proposal, it's just that the preimage itself isn't present there. so I agree it's an annoyance and I'm pretty sure it'll happen less and less in the future, but this has happened in Kusama before (didn't have the courage to check polkascan and find one).

> also is that deposit decided by the size of the call or preimage, or is it a constant, or something else?

Don't take my word for it, @joshua-mir told me afaicr but I might have misinterpreted it.

Reply:
makes sense, thanks for the explanation!